src/igraph_eulerian.c

```diff
@@ -0,0 +1,614 @@
+
+            /* remove edge here */
+            VECTOR(*outgoing_list)[curr]--;
+            VECTOR(*outgoing_list)[next]--;
+            VECTOR(visited_list)[edge] = 1;
+
+            curr = next;
+        } else { /* back track to find remaining circuit */
+            IGRAPH_CHECK(igraph_stack_push(&path, curr));
+            curr = igraph_stack_pop(&tracker);
+            if (!igraph_stack_empty(&edge_tracker)) {
+                curr_e = igraph_stack_pop(&edge_tracker);
+                IGRAPH_CHECK(igraph_stack_push(&edge_path, curr_e));
+            }
+        }
+    }
+
+    while (!igraph_stack_empty(&edge_path)) {
+        IGRAPH_CHECK(igraph_vector_push_back(res, igraph_stack_pop(&edge_path)));           
+    }
+
+    igraph_stack_destroy(&path);
+    igraph_stack_destroy(&tracker);
+    igraph_stack_destroy(&edge_path);
+    igraph_stack_destroy(&edge_tracker);
+    igraph_vector_bool_destroy(&visited_list);
+    igraph_inclist_destroy(&il);
+    IGRAPH_FINALLY_CLEAN(6);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/* Hierholzer algorithm */
+int igraph_i_eulerian_path_undirected(igraph_t *graph, igraph_vector_t *path) {
+
+    /* default starting vertex is 0
+    when we have no odd vertices, we can start anywhere (usually 0)
+    when we have 2 odd vertices, we start at any one of them */
+
+    igraph_integer_t start, i;
+    igraph_inclist_t incl;
+    igraph_vector_int_t *incedges;
+    igraph_vector_t degree, out_degree;
+    igraph_bool_t has_path;
+    igraph_bool_t cycle;
+
+    has_path = 0;
+    cycle = 0;
+
+    IGRAPH_CHECK(igraph_vector_init(&degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &degree);
+    
+    IGRAPH_CHECK(igraph_vector_init(&out_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &out_degree);
+    IGRAPH_CHECK(igraph_degree(graph, &out_degree, igraph_vss_all(), IGRAPH_OUT, IGRAPH_LOOPS));
+
+    IGRAPH_CHECK(igraph_degree(graph, &degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS));
+
+    for (i = 0; i < igraph_vector_size(&degree); i++) {
+        if (((long int) VECTOR(degree)[i]) > 0) {
+            start = i;
+            break;
+        }
+    }
+
+    igraph_vector_destroy(&degree);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &incl, IGRAPH_ALL));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &incl);
+
+    if (cycle && !has_path) {
+        IGRAPH_CHECK(igraph_i_eulerian_path_undirected_implementation(start, graph, path, &out_degree));
+    } else {
+        for (i = 0; i < igraph_vcount(graph); i++) {
+            incedges = igraph_inclist_get(&incl, i);
+            if ( igraph_vector_int_size(incedges) % 2 == 1) {
+                start = i;
+                break;
+            }
+        }
+        igraph_i_eulerian_path_undirected_implementation(start, graph, path, &out_degree);
+    }
+
+    igraph_inclist_destroy(&incl);
+    igraph_vector_destroy(&out_degree);
+    IGRAPH_FINALLY_CLEAN(1);
```

Comment:
This should be 2, not 1.

Please be extremely careful with this because mistakes will lead to hard-to-debug crash. My test program was crashing and it took me quite a while before I managed to find out that it was due to this line. Count the number of `IGRAPH_FINALLY`s and make sure that the match the the `IGRAPH_FINALLY_CLEAN`

Reply:
One trick that we could adopt is that we could check whether the "finally" stack is clean at the end of the unit tests. If it is not, chances are that `IGRAPH_FINALLY_CLEAN()` was invoked with a wrong count somewhere.

See `examples/simple/igraph_strvector.c` for an example; it has something like this at the end:

```
if (!IGRAPH_FINALLY_STACK_EMPTY) {
    /* run in circles, scream and shout */
}
```

Reply:
Excellent catch @szhorvat and idea @ntamas, I'm in favor

Reply:
We could have a macro in the test utilities to help with this and make it a one-liner at the end of every test program.